### PR TITLE
Add centralised build, test & deploy workflow for backend services

### PR DESCRIPTION
**Pull request by samueldobbie**

### CODE REVIEW

**Review by tim-hm — CHANGES_REQUESTED** [review]
Fantastic first iteration! As discussed a summary of next steps/changes:

- [ ] job structure
  
  1. code quality
  2. test & code coverage (tbc - kover)
  3. containerise (needs test)
  4. deploy XYZ (needs containerise, accepts region & cluster-name as env vars)

- [ ] can we set secrets globally for the workflow to be shared between jobs?

- [ ] success criteria
  - can i commit to dev and have it push the container?
  - can i trigger a deploy to alpha?

Heres an example of what I think this needs to look like when finished (and if possible!):

```yaml
name: service-lifecycle

on: push

secrets:
  GH_PAT: ${{ secrets.GH_PAT }}
  AWS_ACCOUNT_ID: ${{ secrets.AWS_ACCOUNT_ID }}
  AWS_ACCESS_KEY_ID: ${{ secrets.AWS_ACCESS_KEY_ID }}
  AWS_SECRET_ACCESS_KEY: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
  SONARQUBE_LOGIN_KEY: ${{ secrets.SONARQUBE_LOGIN_KEY }}

with:
  SERVICE_NAME: vault-attribute
  HELM_SERVICE_NAME: vault-attribute

jobs:
  code-quality-check:
    uses: ...

  test-and-code-coverage:
    uses: ...

  containerise:
    uses: ...

  deploy-dev:
    uses: glimspe-deploy-workflow
    needs: containerise
    with:
      AWS_REGION: eu-west-1
      CLUSTER_NAME: gpsk-alpha-euw1-eks
```

File: .github/workflows/containerise.yml [hunk 1]
```diff
@@ -0,0 +1,151 @@
...
+          token: ${{ secrets.GH_PAT }}
+          path: gradle-plugin
+
+      # Run detekt and tests
+      - name: Set up JDK 16
+        uses: actions/setup-java@v2
+        with:
+          java-version: "17"
+          distribution: "adopt"
+          cache: gradle
```
Comment (tim-hm) [inline .github/workflows/containerise.yml:132]: - please bump the name to JDK 17
- the jdk 17 distribution has changed from adopt to temurin

File: .github/workflows/code-quality-check.yml [hunk 2]
```diff
@@ -0,0 +1,144 @@
...
+          token: ${{ secrets.GH_PAT }}
+          path: gradle-plugin
+
+      # Run detekt and sonarqube
+      - name: Set up JDK 16
+        uses: actions/setup-java@v2
+        with:
+          java-version: "17"
+          distribution: "adopt"
+          cache: gradle
```
Comment (tim-hm) [inline .github/workflows/code-quality-check.yml:130]: Same comment elsewhere about jdk 17

File: .github/workflows/code-quality-check.yml [hunk 3]
```diff
@@ -0,0 +1,144 @@
...
+
+      - name: Grant execute permission for gradlew
+        working-directory: ${{ inputs.SERVICE_NAME }}
+        run: chmod +x gradlew
+
+      - name: Run detekt
+        working-directory: ${{ inputs.SERVICE_NAME }}
+        run: ./gradlew classes detekt
```
Comment (tim-hm) [inline .github/workflows/code-quality-check.yml:138]: Do we need to run `classes` before `detekt`?

File: .github/workflows/code-quality-check.yml [hunk 4]
```diff
@@ -0,0 +1,144 @@
...
+
+      - name: Run sonarqube
+        working-directory: ${{ inputs.SERVICE_NAME }}
+        env:
+          SONARQUBE_LOGIN_KEY: ${{ secrets.SONARQUBE_LOGIN_KEY }}
+        run: ./gradlew sonarqube -Dkotest.tags="!Integration"
```
Comment (tim-hm) [inline .github/workflows/code-quality-check.yml:144]: We should be able to run integration tests now 🤞

File: .github/workflows/code-quality-check.yml [hunk 5]
```diff
@@ -0,0 +1,144 @@
...
+          if [[ ${{ steps.extract-branch.outputs.branch }} == "main" ]]; then
+            COMMONS_BRANCH=main
+          elif [[ '${{ steps.get-branch-commons.outputs.status }}' = '200' ]]; then
+            COMMONS_BRANCH="${{ steps.extract-branch.outputs.branch }}"
+          else
+            COMMONS_BRANCH=dev
+          fi
+
+          echo "Commons branch to checkout: $COMMONS_BRANCH"
+          echo "COMMONS_BRANCH=$COMMONS_BRANCH" >> $GITHUB_ENV
```
Comment (tim-hm) [inline .github/workflows/code-quality-check.yml:52]: Anyway to extract this step into a separate workflow as its used quite regularly?

File: .github/workflows/containerise.yml [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+      - name: Grant execute permission for gradlew
+        working-directory: ${{ inputs.SERVICE_NAME }}
+        run: chmod +x gradlew
+
+      - name: Configure AWS credentials
+        uses: aws-actions/configure-aws-credentials@v1
+        with:
+          aws-access-key-id: ${{ secrets.AWS_ACCESS_KEY_ID }}
+          aws-secret-access-key: ${{ secrets.AWS_SECRET_ACCESS_KEY }}
+          aws-region: eu-west-1
```
Comment (tim-hm) [inline .github/workflows/containerise.yml:143]: Region should come from an env var 🙏

File: .github/workflows/test-and-code-coverage.yml [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+          token: ${{ secrets.GH_PAT }}
+          path: gradle-plugin
+
+      # Run tests
+      - name: Set up JDK 16
+        uses: actions/setup-java@v2
+        with:
+          java-version: "17"
+          distribution: "adopt"
+          cache: gradle
```
Comment (tim-hm) [inline .github/workflows/test-and-code-coverage.yml:128]: JDK 17 comment

**Review by tim-hm — CHANGES_REQUESTED** [review]
Nearly there!!

File: .github/workflows/code-quality-check.yml (hunk 5 above)
Comment (samueldobbie) [reply .github/workflows/code-quality-check.yml]: Yeah, we can have something like the below, although we'd have to move the workflow call to occur within the service (a called workflow can't call another).

```yaml
name: checkout-service

on:
  workflow_call:
    inputs:
      SERVICE_NAME:
        required: true
        type: string
      TARGET_BRANCH:
        required: true
        type: string

    secrets:
      GH_PAT:
        required: true

jobs:
  checkout-branch:
    name: Checkout service @ target branch
    runs-on: ubuntu-latest
    steps:
      - name: Checkout calling service
        uses: actions/checkout@v2
        with:
          path: ${{ inputs.SERVICE_NAME }}

      - name: Extract branch name
        run: echo "::set-output name=branch::$(echo ${GITHUB_REF#refs/heads/})"
        id: extract-branch

      - name: Check for matching branch
        uses: octokit/request-action@v2.x
        id: get-branch-active
        with:
          route: GET /repos/glimpse-protocol/${{ inputs.SERVICE_NAME }}/branches/${{ steps.extract-branch.outputs.branch }}
        env:
          GITHUB_TOKEN: ${{ secrets.GH_PAT }}
        continue-on-error: true

      - name: Determine which ${{ inputs.SERVICE_NAME }} branch to checkout
        run: |
          if [[ ${{ steps.extract-branch.outputs.branch }} == "main" ]]; then
            ACTIVE_BRANCH=main
          elif [[ '${{ steps.get-branch-active.outputs.status }}' = '200' ]]; then
            ACTIVE_BRANCH="${{ steps.extract-branch.outputs.branch }}"
          else
            ACTIVE_BRANCH=dev
          fi

          echo "${{ inputs.SERVICE_NAME }} branch to checkout: $ACTIVE_BRANCH"
          echo "ACTIVE_BRANCH=$ACTIVE_BRANCH" >> $GITHUB_ENV

      - name: Checkout ${{ inputs.SERVICE_NAME }}
        uses: actions/checkout@v2
        with:
          repository: glimpse-protocol/${{ inputs.SERVICE_NAME }}
          ref: ${{ env.ACTIVE_BRANCH }}
          token: ${{ secrets.GH_PAT }}
          path: ${{ inputs.SERVICE_NAME }}
```

Comment (tim-hm) [reply .github/workflows/code-quality-check.yml]: Hmm .. so what does that look like inside the service's workflow file then?

Comment (tim-hm) [reply .github/workflows/code-quality-check.yml]: okay that makes sense ... lets pause that and circle back once its all togehter 👍

**Review by tim-hm — APPROVED** [review]

File: .github/workflows/deploy-to-aws.yml [hunk 8]
```diff
@@ -0,0 +1,60 @@
...
+
+      - name: Pull chart
+        run: |
+          aws ecr get-login-password --region ${{ inputs.AWS_REGION }} | helm registry login ${{ inputs.AWS_ACCOUNT_ID }}.dkr.ecr.${{ inputs.AWS_REGION }}.amazonaws.com --username AWS --password-stdin
+          helm pull oci://${{ inputs.AWS_ACCOUNT_ID }}.dkr.ecr.${{ inputs.AWS_REGION }}.amazonaws.com/helm/${{ inputs.SERVICE_NAME }} --untar
+
+      - name: Deploy
+        working-directory: ${{ inputs.SERVICE_NAME }}
+        run: |
+          echo $(aws sts get-caller-identity)
```
Comment (tim-hm) [inline .github/workflows/deploy-to-aws.yml:54]: is this still needed?

File: .github/workflows/containerise.yml [hunk 9]
```diff
@@ -0,0 +1,160 @@
...
+        type: string
+      SERVICE_NAME:
+        required: true
+        type: string
+
+    secrets:
+      AWS_ACCOUNT_ID:
+        required: true
+      AWS_ACCESS_KEY_ID:
+        required: true
```
Comment (tim-hm) [inline .github/workflows/containerise.yml]: Can we move these into input and then env vars from the calling action please?

**Review by tim-hm — CHANGES_REQUESTED** [review]
Two amendments please and then go ahead and merge without another review 🙏

**Pull request merged**